rpki-validator/src/main/java/net/ripe/rpki/validator3/rrdp/RrdpServiceImpl.java

```diff
@@ -259,30 +248,81 @@ private int processDownloadedSnapshot(RpkiRepository rpkiRepository, RpkiReposit
-        final byte[] deltaBody = rrdpClient.getBody(di.getUri());
-        final byte[] deltaHash = Sha256.hash(deltaBody);
-        if (!Arrays.equals(Hex.parse(di.getHash()), deltaHash)) {
-            rrdpMetrics.update(notification.snapshotUri, ErrorCodes.RRDP_WRONG_DELTA_HASH);
-            throw new RrdpException(ErrorCodes.RRDP_WRONG_DELTA_HASH, "Hash of the delta file " + di + " is " + Hex.format(deltaHash) +
-                    ", but notification file says " + di.getHash());
-        }
+    private void processDelta(RpkiRepository rpkiRepository, RpkiRepositoryValidationRun validationRun, Notification notification, DeltaInfo di, AtomicBoolean changedObjects) {
+        rrdpClient.processUsingTemporaryFile(di.getUri(), Hashing.sha256(), (deltaPath, deltaHash) -> {
+            if (!Arrays.equals(Hex.parse(di.getHash()), deltaHash.asBytes())) {
+                rrdpMetrics.update(notification.snapshotUri, ErrorCodes.RRDP_WRONG_DELTA_HASH);
+                throw new RrdpException(ErrorCodes.RRDP_WRONG_DELTA_HASH, "Hash of the delta file " + di + " is " + Hex.format(deltaHash.asBytes()) +
+                        ", but notification file says " + di.getHash());
+            }
+
+            Long timedStoreDelta = Time.timed(() -> {
+                int counter = 0;
+
+                log.debug("Processing RRDP repository {} delta {}, except for manifests", rpkiRepository.getRrdpNotifyUri(), di.getSerial());
+                counter += processDownloadedDelta(rpkiRepository, validationRun, notification, di, deltaPath, NO_MANIFESTS_PREDICATE);
+
+                log.debug("Processing RRDP repository {} delta {}, manifests only", rpkiRepository.getRrdpNotifyUri(), di.getSerial());
+                counter += processDownloadedDelta(rpkiRepository, validationRun, notification, di, deltaPath, ONLY_MANIFESTS_PREDICATE);
+
+                storage.writeTx0(tx -> rpkiRepositories.update(tx, rpkiRepository));
+
+                changedObjects.set(counter > 0);
+
+                log.info("Added, withdrew, or updated locations for {} new objects", counter);
+            });
+            log.info("Storing delta {} time {}ms", rpkiRepository.getRrdpNotifyUri(), timedStoreDelta);
+
+            return null;
+        });
+    }
+
+    private int processDownloadedDelta(RpkiRepository rpkiRepository, RpkiRepositoryValidationRun validationRun, Notification notification, DeltaInfo di, Path deltaPath, Predicate<RepositoryObjectType> typePredicate) {
+        try (InputStream in = new FileInputStream(deltaPath.toFile())) {
+            AtomicInteger counter = new AtomicInteger(0);
+            AtomicInteger pendingObjectsBytes = new AtomicInteger(0);
+            List<DeltaElement> pendingObjects = new ArrayList<>(1000);
+            Runnable commitPendingObjects = () -> {
+                storage.readTx0(tx -> verifyDeltaIsApplicable(tx, pendingObjects));
+                storage.writeTx0(tx -> counter.addAndGet(
```

Comment:
This one, we already decided, should be adjusted in such a way that parallelStream doesn happen inside of writeTx

Reply:
Did that in a separate pull request https://github.com/RIPE-NCC/rpki-validator-3/pull/259